chronicle/src/main/java/net/openhft/chronicle/network/TcpEventHandler.java

```diff
@@ -0,0 +1,146 @@
+        if (!sc.isOpen()) {
+            handler.onEndOfConnection(sessionDetails);
+            throw new InvalidEventHandlerException("Cannot process, socket chanel is not open.");
+        }
+
+        try {
+            int read = inBB.remaining() > 0 ? sc.read(inBB) : 1;
+            if (read < 0) {
+                closeSC();
+
+            } else if (read >= 0) {
+                // inBB.position() where the data has been read() up to.
+                return invokeHandler();
+            }
+        } catch (IOException e) {
+            handleIOE(e);
+        }
+
+        return false;
+    }
+
+    boolean invokeHandler() throws IOException {
+        long start = System.nanoTime();
+        boolean busy = false;
+        long inBBBPos = inBBB.position();
+        // inBBB.readLimit(inBB.position());
+        inBBB.limit(inBB.position());
+        // outBBB.writePosition(outBB.limit());
+        outBBB.position(outBB.limit());
+        handler.process(inBBB, outBBB, sessionDetails);
+
+        // did it write something?
+        // if (outBBB.writePosition() > outBB.limit() || outBBB.writePosition() >= 4) {
+        if (outBBB.position() > outBB.limit() || outBBB.position() >= 4) {
+            // outBB.limit(Maths.toInt32(outBBB.writePosition()));
+            outBB.limit(Maths.toInt(outBBB.position(), "Int %d out of range"));
+            tryWrite();
+            busy |= true;
+        }
+
+        long compact = System.nanoTime();
+        long inBBCompactStart = -1, inBBCompactEnd = -1;
+        // TODO Optimise.
+        // if it read some data compact();
+        // if (inBBB.readPosition() > 0) {
+        if (inBBB.position() > (inBB.limit() / 2)) {
+            // inBB.position((int) inBBB.readPosition());
+            inBB.position((int) inBBB.position());
+            // inBB.limit((int) inBBB.readLimit());
+            inBB.limit((int) inBBB.limit());
+            inBBCompactStart = System.nanoTime();
+            inBB.compact();
+            inBBCompactEnd = System.nanoTime();
+            inBBB.position(0);
+            // inBBB.readLimit(inBB.position());
+            inBBB.limit(inBB.position());
+            busy |= true;
+        } else {
+            busy |= inBBBPos < inBBB.position(); // we have read something, so deemed busy.
+        }
+        return busy;
+    }
+
+    boolean tryWrite() throws IOException {
+        int wrote = sc.write(outBB);
+        if (wrote < 0) {
+            closeSC();
+
+        } else if (wrote > 0) {
+            outBB.compact().flip();
+            // outBBB.writePosition(outBB.limit());
+            outBBB.position(outBB.limit());
+            // outBBB.writeLimit(outBB.capacity());
+            outBBB.limit(outBB.capacity());
+            return true;
+        }
+        return false;
+    }
+
+    void handleIOE(@NotNull IOException e) {
+        if (!(e instanceof ClosedByInterruptException))
+            e.printStackTrace();
```

Comment:
Would be better to log the exception rather than print the stack trace to stderr